### PR TITLE
Add CollectionIterator::map 

### PR DESCRIPTION
**Pull request by tlode**
Adds array_map helper for CollectionIterator

### CODE REVIEW

File: sources/lib/Model/CollectionIterator.php [hunk 1]
```diff
@@ -9,6 +9,7 @@
  */
 namespace PommProject\ModelManager\Model;
 
+use Closure;
```
Comment (sanpii) [inline sources/lib/Model/CollectionIterator.php]: Useless, simply use `\Closure` below.

File: sources/lib/Model/CollectionIterator.php [hunk 2]
```diff
@@ -198,6 +199,15 @@ public function slice($name)
         return $this->convertSlice(parent::slice($name), $name);
     }
 
+    /**
+     * @param Closure $callback
```
Comment (sanpii) [inline sources/lib/Model/CollectionIterator.php]: Useless line, it’s already in function prototype. That’s duplicate information and add a chance to modify one and not the other.

**Comment by chanmix51** [comment]
Is it different from using a filter and then extract the iterator ?

```php
$identity = function($row) { return $row; };
$result->registerFilter($identity)->extract();
```
Would that behave differently thant `$result->map($identity);` ?

**Comment by tlode** [comment]
Using a filter and the extract method would only return row values of type array. I would like to use `map` to easily transform rows or entities in something different, for example value objects.

```php
$valueObjects = $result->map(
   function ($entity) { 
      return new SomeValueObject($entity->getProperty());
   }
);
```

I have such cases a lot and always have to iterate over the result set externally.

**Comment by chanmix51** [comment]
Yes but this array is at the end used to hydrate entities by the iterator. This is the way we can turn json fields into embedded objects by example.

**Comment by tlode** [comment]
Perhaps, I don't understand the purpose of the filter feature. But I think, it is meant for a different use case. 

In some cases I don't need to have flexible entities in the end. I wan't to transform the result set into something different, something independent.

Using `array_map` is a convenient way to achieve that and adding a `map` function in the `CollectionIterator` seemed appropriate. 

It is another question if it could / should be rather added in the `ResultIterator` instead. I am not sure.

**Comment by chanmix51** [comment]
That’s what I felt. If you do not need entities, wouldn’t that be a `ConvertedResultIterator` method instead ?

**Comment by tlode** [comment]
No, I do all the queries through `Model` methods. `ConvertedResultIterator` wouldn't help, since `CollectionIterator` extends from `ResultIterator`.

**Comment by chanmix51** [comment]
`ConvertedResultIterator` always return an array per row or can be expanded as an array of arrays. The CollectionIterator does the same but with entities. Would’t the `CollectionIterator::map` output something else than an array of entities in that case ?

**Comment by tlode** [comment]
Yes, as mentioned earlier it would return an array of anything you like. In my case simple value objects which I don't treat as entities

**Comment by tlode** [comment]
Sorry, I shall not write comments with the phone :-)

**Comment by chanmix51** [comment]
Ok, the feeling I have regarding this is the following: add the `map` method in the `ConvertedResultIterator` maybe using a PHP generator. Break the inheritance with`CollectionIterator` (it was a _bad_ idea) and use the interface developed in pomm-project/Foundation#71 instead. 
What do you think ?

**Pull request closed without merging**